### PR TITLE
[Proxy] Stitch TCP transport connections regardless to proxy fifos

### PR DESCRIPTION
**Pull request by sergeymatov**
Original proxy flow is likely not needed if TCP transport session is
already established.
This commit also adds some amount of metrics to track UPG flows
stitching and number of stitched flows.

### CODE REVIEW

File: upf/upf_proxy_input.c [hunk 1]
```diff
@@ -122,7 +122,8 @@ kill_connection_hard (tcp_connection_t * tc)
 }
 
 static_always_inline u32
-splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
+splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction,
+		       upf_main_t * gtm)
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:126]: For the sake of consistency, `upf_main_t *` should be the first argument

File: upf/upf_proxy_input.c [hunk 2]
```diff
@@ -133,26 +134,48 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
...
+				     vlib_get_thread_index (), 0, 1);
+      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    }
 
   // check fifo, proxy Tx/Rx are connected...
   if (svm_fifo_max_dequeue (s->rx_fifo) != 0 ||
       svm_fifo_max_dequeue (s->tx_fifo) != 0)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    vlib_increment_simple_counter (&gtm->upf_simple_counters
+				   [UPF_FLOW_NOT_STITCHED_FIFOS],
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:177]: `UPF_FLOW_NOT_STITCHED_FIFOS` is a misnomer here b/c we're ignoring the data in FIFOs and stitching the flows anyway. `UPG_FLOWS_STITCHED_DIRTY_FIFOS` perhaps?

File: upf/upf.h [hunk 3]
```diff
@@ -497,9 +497,30 @@ typedef enum
...
+  UPF_FLOW_NOT_STITCHED_NO_RX_OR_TX = 9,
+  UPF_FLOW_NOT_STITCHED_MSS_MISMATCH = 10,
+  UPF_N_COUNTERS = 11,
 } upf_counters_type_t;
 
+#define foreach_upf_counter_name   \
+  _(ASSOC_COUNTER, total_assoc, upf)     \
+  _(SESSIONS_COUNTER, total_sessions, upf)      \
+  _(FLOW_COUNTER, total_flows, upf)             \
+  _(FLOW_STITCHED, total_stitched, upf)         \
```
Comment (ivan4th) [inline upf/upf.h:515]: starting from `FLOW_STITCHED_`, maybe the prefix should be `FLOWS_STITCHED_` instead?

**Review by ivan4th — CHANGES_REQUESTED** [review]
Some minor (semi-)cosmetic issues

File: upf/upf_proxy_input.c [hunk 4]
```diff
@@ -133,26 +134,48 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
   session_t *s;
 
   if (rev->conn_index == ~0)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    {
+      vlib_increment_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_NOT_STITCHED_NO_REV_CONNECTION],
+				     vlib_get_thread_index (), 0, 1);
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:140]: It is incorrect to increase a counter here b/c the state of the flow may change yet, and rev conn may appear. Meanwhile, the counter may be increased multiple times for the same flow. We only should increase the counters when the stitching is either done or completely and finally rejected for the flow.

File: upf/upf_proxy_input.c [hunk 5]
```diff
@@ -133,26 +134,48 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
...
+      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    }
 
   if (flow->dont_splice)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    {
+      vlib_increment_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_NOT_STITCHED_DONT_STITCH],
+				     vlib_get_thread_index (), 0, 1);
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:148]: No need to increase any counter as this will happen on each packet belonging to a flow for stitching has been rejected for any reason

File: upf/upf.h [hunk 6]
```diff
@@ -497,9 +497,30 @@ typedef enum
   UPF_ASSOC_COUNTER = 0,
   UPF_SESSIONS_COUNTER = 1,
   UPF_FLOW_COUNTER = 2,
-  UPF_N_COUNTERS = 3,
+  UPF_FLOW_STITCHED = 3,
+  UPF_FLOW_NOT_STITCHED_FIFOS = 4,
+  UPF_FLOW_NOT_STITCHED_NO_REV_CONNECTION = 5,
```
Comment (ivan4th) [inline upf/upf.h:502]: This counter should not be there, see below

File: upf/upf.h [hunk 7]
```diff
@@ -497,9 +497,30 @@ typedef enum
...
+  UPF_FLOW_NOT_STITCHED_NO_TRANSPORT = 6,
+  UPF_FLOW_NOT_STITCHED_DONT_STITCH = 7,
```
Comment (ivan4th) [inline upf/upf.h:504]: This counter should not be there, see below

**Review by ivan4th — CHANGES_REQUESTED** [review]
Some more serious issues

File: upf/upf_proxy_input.c [hunk 8]
```diff
@@ -166,14 +189,23 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
   ASSERT (tcp_flow_is_valid (tcpTx, flow, FT_REVERSE ^ direction));
 
   if (!tcpRx || !tcpTx)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    {
+      vlib_increment_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_NOT_STITCHED_NO_RX_OR_TX],
+				     vlib_get_thread_index (), 0, 1);
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:195]: Again, no counter needed, as this is not a permanent rejection

File: upf/upf_proxy_input.c [hunk 9]
```diff
@@ -133,26 +134,48 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
...
+      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    }
 
   s = session_get_if_valid (tc->s_index, tc->thread_index);
   if (!s)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    {
+      vlib_increment_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_NOT_STITCHED_INVALID_SESSION],
+				     vlib_get_thread_index (), 0, 1);
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:169]: no counter needed, as this is not a permanent rejection

File: upf/upf_proxy_input.c [hunk 10]
```diff
@@ -133,26 +134,48 @@ splice_tcp_connection (flow_entry_t * flow, flow_direction_t direction)
...
   // lookup connections
   tc =
     transport_get_connection (TRANSPORT_PROTO_TCP, ftc->conn_index,
 			      ftc->thread_index);
   if (!tc)
-    return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
+    {
+      vlib_increment_simple_counter (&gtm->upf_simple_counters
+				     [UPF_FLOW_NOT_STITCHED_NO_TRANSPORT],
+				     vlib_get_thread_index (), 0, 1);
```
Comment (ivan4th) [inline upf/upf_proxy_input.c:160]: no counter needed, as this is not a permanent rejection

File: upf/upf.h [hunk 11]
```diff
@@ -497,9 +497,30 @@ typedef enum
...
+  _(ASSOC_COUNTER, total_assoc, upf)     \
+  _(SESSIONS_COUNTER, total_sessions, upf)      \
+  _(FLOW_COUNTER, total_flows, upf)             \
+  _(FLOW_STITCHED, total_stitched, upf)         \
+  _(FLOW_NOT_STITCHED_FIFOS, not_stitch_fifos, upf)   \
+  _(FLOW_NOT_STITCHED_NO_REV_CONNECTION, no_rev_connection, upf) \
+  _(FLOW_NOT_STITCHED_NO_TRANSPORT, no_transport, upf) \
+  _(FLOW_NOT_STITCHED_DONT_STITCH, dont_stitch, upf) \
+  _(FLOW_NOT_STITCHED_INVALID_SESSION, invalid_session, upf) \
+  _(FLOW_NOT_STITCHED_NO_RX_OR_TX, no_rx_or_tx, upf)    \
```
Comment (ivan4th) [inline upf/upf.h:521]: not needed, see below

**Review by ivan4th — CHANGES_REQUESTED** [review]
Actually, it was my fault, when I was pointing to https://github.com/travelping/upg-vpp/commit/f37066c4c3c9bd1597c5d6bbd9455e39cdd40526 it did not come to my mind that counters need a bit different approach. They should only be changed when the flow is either stitched, or when stitching is completely and finally rejected for it.

**IMPORTANT!!!** You also need to add counters for these "don't splice" cases:
```
  /* check TCP connection properties */
  if ((tcpRx->snd_mss > tcpTx->rcv_opts.mss) ||
      (tcpTx->snd_mss > tcpRx->rcv_opts.mss))
    {
      upf_debug ("=============> DON'T SPLICE <=============");
      flow->dont_splice = 1;
      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
    }

  if (tcp_opts_tstamp (&tcpTx->rcv_opts) !=
      tcp_opts_tstamp (&tcpRx->rcv_opts))
    {
      upf_debug ("=============> DON'T SPLICE <=============");
      flow->dont_splice = 1;
      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
    }

  if (tcp_opts_sack_permitted (&tcpTx->rcv_opts) !=
      tcp_opts_sack_permitted (&tcpRx->rcv_opts))
    {
      upf_debug ("=============> DON'T SPLICE <=============");
      flow->dont_splice = 1;
      return UPF_PROXY_INPUT_NEXT_TCP_INPUT;
    }
```

Basically, only these 3 non-stitching cases and also 2 stitching cases (ok and dirty fifo) is all what should be counted.

**Review by ivan4th — APPROVED** [review]
LGTM
